### ISSUE TITLE
conversation_error, 30 mins of inactivity

### ISSUE THREAD

**Issue by natt-tester** [GluuFederation/oxTrust#1029]
Steps:

1. Log in
2. Click on the menu icon several times
3. Try to change the language to Russian

Expected: The language can be changed.
Actual: There is an error message: "Oops
The form was removed after 30 minutes of inactivity 
Return to the application using below button." 

![2018-07-03_21h54_25](https://user-images.githubusercontent.com/38302725/42218379-b61504b6-7f0b-11e8-9d52-55d363758da9.png)

(I don't have access to the logs.)

**Comment by syntrydy** [comment]
@natt-tester i'm unable to reproduce it. Btw, during my work the page above often shows up. 

Can you provide more context?

**Comment by natt-tester** [comment]
I caught it again and got the logs. It seems quite random, I'm afraid, but maybe the logs will show why this happens:

**Comment by syntrydy** [comment]
I haven't saw such behavior in 3.1.4

**Comment by natt-tester** [comment]
**Turns out this behavior is easily reproducible in RC5:**

1. Log in to the oxTrust
2. Click on a menu item, e.g. UMA
3. Click on an item from the UMA menu, e.g. Scopes
4. When it's loaded, double click on it again.

Tested locally in Ubuntu16 and on c7.
![green_screen](https://user-images.githubusercontent.com/38302725/45757084-e4bb5080-bc65-11e8-9ee1-51285d69fc7e.gif)

**Comment by syntrydy** [comment]
Why do you double click on a link, instead of clicking?

Double click ==> Click * 2 and this lead to some conversation error in application.

**Comment by natt-tester** [comment]
@syntrydy, no GUI behavior should lead to an error like this. We have to make sure our users get the best experience, even if they double-click something by mistake. Let's make it as user-friendly as possible :)

**Comment by yurem** [comment]
I agree about "double click" issue.

But this error message due to user inactivity looks good for me:
```
Actual: There is an error message: "Oops
The form was removed after 30 minutes of inactivity
Return to the application using below button."
```

Why do you think application should allow to do action after 30 minutes of user inactivity? We can't use long sessions because it leds to security errors.

**Comment by syntrydy** [comment]
i didn't found a fix for this issue

**Comment by yurem** [comment]
According to stack trace it's JSF 2.2.16 issue.
In our code when you navigate from one menu action to another one we call method to end current conversation. It works well until you send requests too often. You send few similar requests which do the same. First one is end conversation and second one led to error because conversation context removed already.

**Comment by yurem** [comment]
I think we can change milestone to 4.0.0 to think about it later.

**Comment by yurem** [comment]
BTW, here is how other Shib IDP3 plugin handle conversation timeout issue: https://github.com/GluuFederation/oxShibboleth/blob/version_3.1.4/oxShibbolethWebapp/src/main/webapp/no-conversation-state.jsp#L19
It just show error page with message.

I think we only can improve here error message if needed. Both flow looks good for me:
1) oxTrust has conversation lifetime = session lifetime = 30 minutes. Hence user might get session timeout or conversation timeout which depends on use case
2) When you click multiple time on link... First request end conversation, than application break processing request due to second one. And as result second request can't load conversation because first one ended it already. I'm not sure that we need to fix this. What do you think?

**Comment by yurem** [comment]
We need to revert commit after resolving this issue: https://github.com/GluuFederation/oxTrust/commit/a0810408bbc024f38eafea043559c0486e1948d3

This is possible solution to fix this issue: https://www.knitelius.com/2015/05/10/synchronizing-access-to-cdi-conversations/

The main idea of it to allow access one conversation only one request. It's introduce synchronization

**Comment by syntrydy** [comment]
Fixed

**Comment by natt-tester** [comment]
I'm unable to reproduce this behavior in 3.1.5_3. It's DONE.

**Comment by natt-tester** [comment]
I've just reproduced this behavior in Export/Import Ldif view:

![screencast 2018-12-18 06-01-41](https://user-images.githubusercontent.com/38302725/50133222-edceb480-028a-11e9-97e3-123baf77c509.gif)

**Comment by syntrydy** [comment]
@natt-tester  Better create a separate issue for it.

**Comment by natt-tester** [comment]
@syntrydy it's reopened in #1387 with an updated description.

**Comment by natt-tester** [comment]
@syntrydy I think we need to keep this issue open to group all the misbehaving elements.
I've just confirmed the fix for #1387, but the noticed the same behavior while clicking fast on Cancel in **Cache Refresh.**

Tested in v.3.1.5_4

**Comment by syntrydy** [comment]
@natt-tester if we encounter this error somewhere else we can open a new issue
